### PR TITLE
fix(scripts/Naxx): The art of Naxxramas (1/2)

### PR DESCRIPTION
**Pull request by Si1ker**
image

<Silker> As the biggest fan of Naxxramas, it is time to give my work instead of letting it waste. Much love to **Naxxramas**.

This Pull Request aims to import all the private fixes for the Naxxramas raid I did from my closed project "Naxxathar".
It is an improvement/overhaul for this raid at cpp level and it will set it almost blizzlike, it took me nearly 1 year to reach this point of perfection, the second part will solve all minor issues and do the final adjustments.

AzerothCore have half of the issues/bugs reported here in 2014: https://github.com/dalaranwow/dalaran-wow/issues/1415

NOTE: Damage output will be reviewed and adjusted in #4195 

The whole work will be split into two PRs:

Part 1 (this): Will solve mostly script issues & broken mechanics = (scripts)
Part 2: Will solve all issues regarding objects, environment, NPCs and SAI = (database)

### = = = = Sources = = = =
The sources for the following changes are a combination of TankSpot videos, Naxx 2008/2009 runs on YouTube and online guides as well, most of the time, the online guides were wrong so on specific cases I had to confirm with videos.
[TankSpot Naxxramas - Playlist](https://www.youtube.com/watch?v=g47oN5snpSU&list=PL602E26F552DBF4BE)
[WoWwiki Naxxramas raid guide](https://wowwiki-archive.fandom.com/wiki/Naxxramas)
[Naxxramas videos on YT filtered by time posted, look for the bottom ones](https://www.youtube.com/results?search_query=naxxramas+full+clear&sp=CAI%253D)
[Naxxramas videos from WoWAholics101](https://www.youtube.com/results?search_query=naxxramas+WoWAholics101)
### = = = = = = = = = = = =

"Imagine an open-source having better raid-quality than top private servers in the world" -Silker

## Changes Proposed:

image
### Anub'Rekhan:
- Refactor
- Include emotes for the Crypt Guard (on dead and spawn)
- Add Crypt Guard spawn event in 10man difficulty
- Adjust Crypt Guard initial locations in 25man difficulty
- Change spawn locations for "Crypt Guard"
- Improve target selection for "Corpse Scarab"
- Fix enrage smart_script for "Crypt Guard"
- Add enrage emote for "Crypt Guard"
- Adjust initial cast for "Locust Swarm"
- Add 'CallForHelp' check if minions get in combat prior to Anub'Rekhan

image
### Grand Widow Faerlina:
- Refactor
- Fix entrance-web at engage/dead/reset
- Adjust available targets for "Poison volley" in 25man/10man raid difficulty
- Adjust timers for "Poison bolt volley"
- Adjust timers for "Rain of fire"
- Add 'CallForHelp' to check if minions get in combat prior to Faerlina
- Add Widow's Embrace (the minion version of the spell)
- Kill the Naxxramas Worshipper in 25mode when is forced to use Widow's Embrace (the minion version of the spell)
- Avoid cast natural spells and execute emote when Faerlina is under the effects of Widow's Embrace in 25mode

image
### Maexxna:
- Refactor
- Adjust timer for "Poison shock"
- Despawn all summons at death

image
### Noth the Plaguebringer:
- Refactor
- Replace 'NearTeleportTo' with the proper spell for that action
- Fix door entrance
- Avoid reset position at death if is already at ground
- Adjust room limits before boss reset so the door can work properly
- Add "Berserk" spell
- Improve spawning positions
- Adjust timers for the initial summoning event and between summoning events

(credits to @Nefertumm for the following four changes)
- Add "times in balcony" count
- Fix Balcony Summon types and structure
- Add 'Berserk mechanic' = Noth will Berserker Enrage after three skeleton phases.
- Remove phase events

image
### Heigan the Unclean:
- Refactor
- Enable random movement
- Change initial orientation
- Adjust initial timer for "Spell Disruption"
- Adjust the initial timer for "Decrepit Fever"
- Adjust timers between each "Decrepit Fever"
- Adjust timer for initial "Eruption" at 'Slow dance' and 'Fast dance'
- Add spell "Teleport Self"
- Adjust orientation at the start of Phase 2
- Delay "Plague cloud" after Heigan teleports to the platform
- (Fix): SAY_DEATH  ->  EMOTE_DEATH

image
### Loatheb:
- Refactor
- Fix event and timer for "Inevitable doom"
- Fix non-working emotes by adding additional events for "Necrotic aura" warnings
- Delete spell_script for "Necrotic Aura Warning"
- Adjust timer for initial 'summon spore'
- Adjust timer for initial "Deathbloom"
- Adjust movement speed for Spores

image
### Instructor Razuvious:
- Refactor
- Adjust timers for "Disrupting Shout"
- Adjust timers for "Jagged Knife"
- Adjust timers for "Unbalancing strike"
- Change initial positions for Death Knight Understudies
- Switch "Disrupting Shout" spells in 10/25 mode
- Fix and refactor the system/structure for all emotes / texts / dialogues / sound files / chance / group
- Disable Death Knight Understudies abilities usage on their own

image
### Gothik the Harvester:
- Refactor
- Remove NOT_SELECTABLE flags
- Delete initial position change after the first attack received
- Add teleport spells
- Replace teleport methods with actual spells built-in the core
- Adjust timer for "Shadow bolt"
- Refactor spell/event tables
- Add minion spells on 25mode
- Remove "Unholy aura" from the script
- Fix EVENT_SPELL_STOMP for 'dead horse'
- Remove empty lines
- Adjust timers for minion spells
- Add additional checks for when a minion can use its ability to avoid range miss
- Fixed dot stack of "Death Plague" spell in #4299 
- Add missing aura  "Unholy aura" to NPC "Spectral raider"

image
### Four Horsemen:
- Refactor
- Fix marks damage
- Add emote for Unyielding Pain/Condemnation
- Adjust timers for Unyielding Pain/Condemnation
- Adjust berserk timer

image
### Patchwerk:
- Refactor
- Adjust the initial timer for "Hateful Strike"
- Adjust the initial timer for "Slime Bolt"

image
### Grobbulus:
- Refactor
- Change emote for 'slime spray' to a boss emote
- Fix and implement script for Mutating Injection (core mechanic)
- Adjust timers

image
### Gluth:
- Refactor
- Adjust timer for berserbk
- Divide decimate timer for 10man mode and 25man mode

image
### Thaddius:
- Refactor
- Replace conditions
- Add missing game object for the second tesla coil
- Adjust timers for Stalagg/Feugeun transition to Thaddius encounter
- Re-do disables for tesla visuals
- Create "Thaddius opening" event
- Remove useless code StartEvent()
- Fix room door
- Adjust Thaddius position after he's able to use "ball lightning"
- Rename some events for better recognition
- Set invisible triggers in their position to avoid chase players
- Adjust target selection for "ball lightning"

image
### Sapphiron:
- Refactor
- Add missing emotes
- Delay door opening

image
### Kel'Thuzad:
- Refactor
- Add minions at the gate for phase 1
- Activate portals for phase 3
- Add initial channeling at phase 1
- Adjust CC immunity for Guardian of Icecrown
- Fix Enrage
- Randomize initial minions movement at phase 1
- Implement spawn pools
- Adjust timers for spells & summons at phase 1 to met blizzlike amount
- Fix emote at the beginning for phase 2
- Create "flee" behavior for Guardians of Icecrown after the Kel'Thuzad's death
- Design summoning system for phase 1 summons using SummonHelper();

### = = Other = =
- Removed incorrect [Double Attack](https://db.rising-gods.de/?spell=18943) in #4773 for [Venom Stalkers](https://www.wowhead.com/npc=15976/venom-stalker)

## Issues Addressed:
- Closes around +100 issues never reported

## Tests Performed:
- These were the scripts I used in my private server "Naxxathar" when it was live, were executed week by week with different raids
- Everything was tested in the dev branch then re-tested in closed betas before implementation in live branch/server

## How to Test the Changes:

Run Naxxramas & check/compare if the changes proposed are met

## Known Issues: I was not able to fix the following issues in this Pull Request:
- #4139 
- #4140 
- #4198 
- #4273 
- #4663 
- #4923 

## TODO List:

- [x] Anub'Rehkan
- [x] Faerlina
- [x] Maexxna
- [x] Noth
- [x] Heigan
- [x] Loatheb
- [x] Razuvious
- [x] Gothik
- [x] Horsemen
- [x] Patchwerk
- [x] Grobbulus
- [x] Gluth
- [x] Thaddius
- [x] Sapphiron
- [x] Kel'Thuzad
= PR part 1 completed on 3/22/20 =

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Nefertumm** [comment]
Ty for credits. Nice pr

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp [hunk 1]
```diff
@@ -182,7 +182,7 @@ class boss_four_horsemen : public CreatureScript
             events.Reset();
             // Schedule Events
             events.RescheduleEvent(EVENT_SPELL_MARK_CAST, 24000);
-            events.RescheduleEvent(EVENT_BERSERK, 100 * 15000);
+            events.RescheduleEvent(EVENT_BERSERK, 600000);
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp:185]: what about using constants like in other places? such as `5 * MINUTES` ? so it's more readable

Comment (Si1ker) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp:185]: we work in milliseconds, check 99% of scripts.

**Comment by FrancescoBorzi** [comment]
seems all green now @Si1ker , good job!

**Review by 55Honey — COMMENTED** [review]
SQL checked. Can't help much with 
C++ tho. No malicicous Code noticed anywhere. Formating seems proper to me.

**Review by 55Honey — DISMISSED** [review]

**Comment by iyasenai** [comment]
Tested.

**Comment by claudiodfc** [comment]
Tested. Better than before. Can be merged.

**Comment by Dismishal** [comment]
Working as suggested!
Tested based on changes applied.

**Comment by Delcidx** [comment]
Tested in game, working. :smirk_cat:

**Comment by RemiliaSc** [comment]
Tested with a friend, amazing work!
Thanks for the PR, I've been waiting for it  n.n

**Comment by Kenjbr** [comment]
Tested locally while checking changes, totally improved!

**Comment by MathematicalWTF** [comment]
Tested.
The best I have seen so far.

**Comment by Si1ker** [comment]
Thanks everyone for testing, much appreciated! 🚀

**Review by claudiodfc — APPROVED** [review]

File: src/server/scripts/Northrend/Naxxramas/boss_loatheb.cpp [hunk 2]
```diff
@@ -126,24 +140,39 @@ class boss_loatheb : public CreatureScript
...
-                    events.RepeatEvent(events.GetTimer() < 5 * MINUTE * IN_MILLISECONDS ? 30000 : 15000);
+                    doomCounter++;
+                    if (doomCounter < 6)
+                    {
+                        events.RepeatEvent(30000);
+                    }
+                    else
+                    {
+                        events.RepeatEvent(15000);
+                    }
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_loatheb.cpp:164]:
```suggestion
                    events.RepeatEvent(doomCounter < 6 ? 30000 : 15000);
```
not sure what is the best code style

File: src/server/scripts/Northrend/Naxxramas/boss_noth.cpp [hunk 3]
```diff
@@ -219,28 +243,36 @@ class boss_noth : public CreatureScript
...
-                        SummonHelper(NPC_PLAGUED_CHAMPION, RAID_MODE(2, 4));
-                    else if (events.GetPhaseMask() == 1)
+                    me->CastSpell(me, SPELL_SUMMON_PLAGUED_WARRIORS, true); // visual
+                    switch (timesInBalcony)
                     {
-                        SummonHelper(NPC_PLAGUED_CHAMPION, RAID_MODE(1, 2));
-                        SummonHelper(NPC_PLAGUED_GUARDIAN, RAID_MODE(1, 2));
+                         case 0:
+                             SummonHelper(NPC_PLAGUED_CHAMPION, RAID_MODE(2, 4));
+                             break;
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_noth.cpp:255]: this is useless because it is the same to the default condition

Comment (Helias) [reply src/server/scripts/Northrend/Naxxramas/boss_noth.cpp:255]: so if / else was good here

File: src/server/scripts/Northrend/Naxxramas/boss_anubrekhan.cpp [hunk 4]
```diff
@@ -158,12 +176,26 @@ class boss_anubrekhan : public CreatureScript
...
             }
-
             ScriptedAI::MoveInLineOfSight(who);
         }
 
         void UpdateAI(uint32 diff) override
         {
+            if (!me->IsInCombat() && sayGreet)
+            {
+                for( std::list<uint64>::iterator itr = summons.begin(); itr != summons.end(); ++itr )
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_anubrekhan.cpp:186]: you can make it shorter with `auto itr = .....`

File: src/server/scripts/Northrend/Naxxramas/boss_razuvious.cpp [hunk 5]
```diff
@@ -88,40 +85,50 @@ class boss_razuvious : public CreatureScript
...
-            if (pInstance)
-                pInstance->SetData(DATA_IMMORTAL_FAIL, 0);
+            if (roll_chance_i(30))
+            {
+                Talk(SAY_SLAY);
+            }
+            if (who->GetTypeId() == TYPEID_PLAYER)
+            {
+                if (pInstance)
+                {
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_razuvious.cpp:95]: merge these 2 if conditions

File: src/server/scripts/Northrend/Naxxramas/boss_anubrekhan.cpp [hunk 6]
```diff
@@ -158,12 +176,26 @@ class boss_anubrekhan : public CreatureScript
...
+                {
+                    if (pInstance)
+                    {
+                        if (Creature* cr = pInstance->instance->GetCreature(*itr))
+                        {
+                            if (cr->IsInCombat())
+                                DoZoneInCombat();
+                        }
+                    }
+                }
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_anubrekhan.cpp:196]:
```suggestoin
for(auto itr = summons.begin(); itr != summons.end(); ++itr )
                {
                    if (pInstance)
                    {
                        if (auto cr = pInstance->instance->GetCreature(*itr) && cr->IsInCombat())
                        {
                            DoZoneInCombat();
                        }
                    }
                }
```

File: src/server/scripts/Northrend/Naxxramas/boss_razuvious.cpp [hunk 7]
```diff
@@ -185,11 +187,13 @@ class boss_razuvious_minion : public CreatureScript
...
         {
-            if (who->GetTypeId() != TYPEID_PLAYER)
-                return;
-
-            if (me->GetInstanceScript())
-                me->GetInstanceScript()->SetData(DATA_IMMORTAL_FAIL, 0);
+            if (who->GetTypeId() == TYPEID_PLAYER)
+            {
+                if (me->GetInstanceScript())
+                {
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_razuvious.cpp:193]: merge these 2 conditions

File: src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp [hunk 8]
```diff
@@ -71,23 +74,28 @@ class boss_faerlina : public CreatureScript
...
+            summons.Summon(cr);
+        }
 
         void Reset() override
         {
             BossAI::Reset();
             events.Reset();
             summons.DespawnAll();
             SummonHelpers();
+            if (GameObject* go = me->FindNearestGameObject(SPIDER_WEB_ENTRANCE, 100.0f))
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp:95]: @Si1ker not necessary, just a tip: can be shorter using `auto go` instead of having to specify the type

File: src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp [hunk 9]
```diff
@@ -258,45 +276,70 @@ class boss_thaddius : public CreatureScript
...
+                        }
+                    }
+                    if (GameObject* go = me->FindNearestGameObject(GO_TESLA_COIL_LEFT, 100.0f))
+                    {
+                        go->SetGoState(GO_STATE_READY);
+                    }
+                    if (GameObject* go = me->FindNearestGameObject(GO_TESLA_COIL_RIGHT, 100.0f))
+                    {
+                        go->SetGoState(GO_STATE_READY);
+                    }
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp:300]: this kind of if() is repeated a lot of times, you could create an external function and re-use it in all this PR

File: src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp [hunk 10]
```diff
@@ -119,20 +130,42 @@ class boss_faerlina : public CreatureScript
...
+                    if (pInstance)
+                    {
+                        if (Creature* cr = pInstance->instance->GetCreature(*itr))
+                        {
+                            if (cr->IsInCombat())
+                                DoZoneInCombat();
+                        }
+                    }
+                }
+            }
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp:167]: same code suggestion as above, as the code is the same. We could also consider re-using this code somehow

File: src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp [hunk 11]
```diff
@@ -537,20 +594,29 @@ class spell_thaddius_pos_neg_charge : public SpellScriptLoader
         {
             uint8 count = 0;
             for (auto& ihit : targets)
+            {
                 if (ihit->GetGUID() != GetCaster()->GetGUID())
+                {
                     if (Player* target = ihit->ToPlayer())
+                    {
                         if (target->HasAura(GetTriggeringSpell()->Id))
+                        {
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp:603]: you could merge these conditions

File: src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp [hunk 12]
```diff
@@ -537,20 +594,29 @@ class spell_thaddius_pos_neg_charge : public SpellScriptLoader
...
-
                 if (GetSpellInfo()->Id == SPELL_POSITIVE_CHARGE)
+                {
                     spellId = SPELL_POSITIVE_CHARGE_STACK;
+                }
                 else // if (GetSpellInfo()->Id == SPELL_NEGATIVE_CHARGE)
+                {
                     spellId = SPELL_NEGATIVE_CHARGE_STACK;
-
+                }
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp:619]: `spellId = GetSpellInfo()->Id == SPELL_POSITIVE_CHARGE ? SPELL_NEGATIVE_CHARGE_STACK : SPELL_NEGATIVE_CHARGE_STACK;`

File: src/server/scripts/Northrend/Naxxramas/naxxramas.h [hunk 13]
```diff
@@ -78,19 +82,20 @@ enum NXGOs
...
+    GO_KELTHUZAD_PORTAL_4           = 181405,
 
-    GO_DEATHKNIGHT_WING             = 181577, //Loatheb portal
-    GO_THADDIUS_PORTAL              = 181576, //Thadius portal
-    GO_MAEXXNA_PORTAL               = 181575, //Maexxna portal
-    GO_HORSEMAN_PORTAL              = 181578, //Four Horseman portal
+    GO_DEATHKNIGHT_WING             = 181577,
+    GO_THADDIUS_PORTAL              = 181576,
+    GO_MAEXXNA_PORTAL               = 181575,
+    GO_HORSEMAN_PORTAL              = 181578,
```
Comment (Helias) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.h:93]: why did you remove the comments?

**Review by Helias — CHANGES_REQUESTED** [review]

File: src/server/scripts/Northrend/Naxxramas/boss_loatheb.cpp (hunk 2 above)
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_loatheb.cpp:164]:
```suggestion
events.RepeatEvent(doomCounter < 6 ? 30000 : 15000)
```

File: src/server/scripts/Northrend/Naxxramas/boss_patchwerk.cpp [hunk 14]
```diff
@@ -21,20 +21,20 @@ enum Spells
...
-    EVENT_SPELL_BERSERK             = 4,
+    EVENT_HATEFUL_STRIKE            = 2,
+    EVENT_SLIME_BOLT                = 3,
+    EVENT_BERSERK                   = 4
 };
 
 enum Misc
 {
-    ACHIEV_TIMED_START_EVENT        = 10286,
+    ACHIEV_TIMED_START_EVENT        = 10286
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_patchwerk.cpp:37]: why? it's ok to keep the comma always, so if you add a new row you don't need to change the existing ones ;)

File: src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp [hunk 15]
```diff
@@ -258,45 +276,70 @@ class boss_thaddius : public CreatureScript
...
+                    {
+                        if (Creature* cr = ObjectAccessor::GetCreature(*me, (*itr)))
+                        {
+                            if (cr->GetEntry() == NPC_TESLA_COIL)
+                            {
+                                Unit::Kill(cr, cr);
+                            }
+                        }
+                    }
+                    if (GameObject* go = me->FindNearestGameObject(GO_TESLA_COIL_LEFT, 100.0f))
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp:293]:
```suggestion
                    for (auto itr = summons.begin(); itr != summons.end(); ++itr)
                    {
                        if (auto cr = ObjectAccessor::GetCreature(*me, (*itr)) && cr->GetEntry() == NPC_TESLA_COIL)
                        {
                            Unit::Kill(cr, cr);
                        }
                    }
```

File: src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp [hunk 16]
```diff
@@ -119,20 +130,42 @@ class boss_faerlina : public CreatureScript
...
+                for( std::list<uint64>::iterator itr = summons.begin(); itr != summons.end(); ++itr )
+                {
+                    if (pInstance)
+                    {
+                        if (Creature* cr = pInstance->instance->GetCreature(*itr))
+                        {
+                            if (cr->IsInCombat())
+                                DoZoneInCombat();
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (claudiodfc) [reply src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp:165]: why is DoZoneInCombat() even necessary?

**Review by FrancescoBorzi — APPROVED** [review]

File: src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp (hunk 9 above)
Comment (claudiodfc) [reply src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp:300]: maybe something like creatureAI->SetGoState(GO_TESLA_COIL_RIGHT, GO_TESLA_COIL_RIGHT);

And then inside creatureAI search for the gameobject in 5000.0f

It's not optimal but it's reusable

Comment (claudiodfc) [reply src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp:300]: I guess what's best should be creating a Gameobject pointer list/vector that would be initiated on create. Then search for the uint32 id passed onto a function and just activate whatever object necessary.

File: src/server/scripts/Northrend/Naxxramas/naxxramas.h (hunk 13 above)
Comment (claudiodfc) [reply src/server/scripts/Northrend/Naxxramas/naxxramas.h:93]: Redundant commentaries, it was already defined as "GO_BOSS_PORTAL" so no point in having a commentary saying // boss_portal

File: src/server/scripts/Northrend/Naxxramas/boss_patchwerk.cpp (hunk 14 above)
Comment (claudiodfc) [reply src/server/scripts/Northrend/Naxxramas/boss_patchwerk.cpp:37]: I would go further ahead and remove the enum completely since an abstracted user-defined type called `Misc` is so unnecessary. This should instead be a `const uint32 `. Assuming the script for Patchwerk is perfect this enum is not needed.
